examples/viewer/viewer.cc

```diff
@@ -190,10 +192,91 @@ static void CalcNormal(float N[3], float v0[3], float v1[3], float v2[3]) {
     float len = sqrtf(len2);
 
     N[0] /= len;
-    N[1] /= len;
+	N[1] /= len;
+	N[2] /= len;
   }
 }
 
+namespace // Local utility functions
+{
+	void addBtoA(float a[3], float b[3])
+	{
+		for (size_t i = 0; i < 3; ++i)
+			a[i] += b[i];
+	}
+
+	void assignBtoA(float a[3], float b[3])
+	{
+		for (size_t i = 0; i < 3; ++i)
+			a[i] = b[i];
+	}
+
+	void normalizeVector(float N[3])
+	{
+		float len2 = N[0] * N[0] + N[1] * N[1] + N[2] * N[2];
+		if (len2 > 0.0f) {
+			float len = sqrtf(len2);
+
+			N[0] /= len;
+			N[1] /= len;
+			N[2] /= len;
+		}
+	}
+
+	void computeSmoothingNormals(tinyobj::attrib_t &attrib, tinyobj::shape_t &shape,
+		std::map<int, float[3]>& smoothVertexNormals)
+	{
+		smoothVertexNormals.clear();
+		std::map<int, float[3]>::iterator iter;
+
+		for (size_t f = 0; f < shape.mesh.indices.size() / 3; f++) {
```

Comment:
The idea: 
(1) For each face, compute its normal and that to each of its vertices.
(2) Then normalize (unit vector) the summed normal at each vertex.